core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java

```diff
@@ -0,0 +1,102 @@
+package org.optaplanner.constraint.streams.bavet.common;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import org.optaplanner.constraint.streams.bavet.uni.UniTuple;
+import org.optaplanner.constraint.streams.bavet.uni.UniTupleImpl;
+
+public abstract class AbstractMapNode<InTuple_ extends Tuple, Right_>
+        extends AbstractNode
+        implements TupleLifecycle<InTuple_> {
+
+    private final int inputStoreIndex;
+    /**
+     * Calls for example {@link AbstractScorer#insert(Tuple)} and/or ...
+     */
+    private final TupleLifecycle<UniTuple<Right_>> nextNodesTupleLifecycle;
+    private final int outputStoreSize;
+    private final Queue<UniTuple<Right_>> dirtyTupleQueue;
+
+    protected AbstractMapNode(int inputStoreIndex, TupleLifecycle<UniTuple<Right_>> nextNodesTupleLifecycle,
+            int outputStoreSize) {
+        this.inputStoreIndex = inputStoreIndex;
+        this.nextNodesTupleLifecycle = nextNodesTupleLifecycle;
+        this.outputStoreSize = outputStoreSize;
+        dirtyTupleQueue = new ArrayDeque<>(1000);
+    }
+
+    @Override
+    public void insert(InTuple_ tuple) {
+        if (tuple.getStore(inputStoreIndex) != null) {
+            throw new IllegalStateException("Impossible state: the input for the tuple (" + tuple
+                    + ") was already added in the tupleStore.");
+        }
+        Right_ mapped = map(tuple);
+        UniTuple<Right_> outTuple = new UniTupleImpl<>(mapped, outputStoreSize);
+        tuple.setStore(inputStoreIndex, outTuple);
+        dirtyTupleQueue.add(outTuple);
+    }
+
+    protected abstract Right_ map(InTuple_ tuple);
+
+    @Override
+    public void update(InTuple_ tuple) {
+        UniTupleImpl<Right_> outTuple = tuple.getStore(inputStoreIndex);
+        if (outTuple == null) {
+            // No fail fast if null because we don't track which tuples made it through the filter predicate(s)
+            insert(tuple);
```

Comment:
Why do we insert the tuple in this case?

Reply:
This is a bit of a quirk of the filter implementation. Insert can be totally filtered out, they will not propagate. Which means that if we're seeing the tuple for the first time doing update, it actually means an insert.

Reply:
Ok. Maybe worth mentioning in the comment (or perhaps it's clear for those more familiar with Bavet)?

Reply:
Being familiar with Bavet, the existing comment makes sense to me.
That said, there should eventually be a more general documentation for Bavet which would explain the inner workings - this being one of them.